File: languageExtensions/OrchidDiagrams/src/main/java/com/eden/orchid/languages/diagrams/PlantUmlCompiler.java
package com.eden.orchid.languages.diagrams;

import com.eden.orchid.api.compilers.OrchidCompiler;
import net.sourceforge.plantuml.FileFormat;
import net.sourceforge.plantuml.FileFormatOption;
import net.sourceforge.plantuml.SourceStringReader;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.Map;

@Singleton
public class PlantUmlCompiler extends OrchidCompiler {

    @Inject
    public PlantUmlCompiler() {
        super(800);
    }

    @Override
    public String compile(String extension, String source, Map<String, Object> data) {
        try {
            try {
                // ensure string is wrapped in @startuml...@enduml
                source = source.trim();
                if(!source.startsWith("@startuml")) {
                    source = "@startuml\n" + source;
                }
                if(!source.endsWith("@enduml")) {
                    source = source + "\n@enduml";
                }

                // compile string to SVG
                ByteArrayOutputStream os = new ByteArrayOutputStream(1024);

                FileFormatOption fileFormat = new FileFormatOption(FileFormat.SVG);
                fileFormat.hideMetadata();

                SourceStringReader reader = new SourceStringReader(source);
                reader.outputImage(os, fileFormat);
                os.close();

                String s = new String(os.toByteArray(), Charset.forName("UTF-8"));
                s = s.replaceAll("<\\?(.*)\\?>", ""); // remove XML declaration

                return s;
            }
            catch (IOException e) {
                e.printStackTrace();
                return "";
            }
        }
        catch (Throwable e) {
            e.printStackTrace();
            return "";
        }
    }

    @Override
    public String getOutputExtension() {
        return "svg";
    }

    @Override
    public String[] getSourceExtensions() {
        return new String[]{"uml"};
    }
}

File: languageExtensions/OrchidDiagrams/src/main/kotlin/com/eden/orchid/languages/diagrams/PlantUmlCompiler.kt
package com.eden.orchid.languages.diagrams

import com.eden.orchid.api.compilers.OrchidCompiler
import net.sourceforge.plantuml.FileFormat
import net.sourceforge.plantuml.FileFormatOption
import net.sourceforge.plantuml.SourceStringReader
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.charset.Charset
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PlantUmlCompiler @Inject
constructor() : OrchidCompiler(800) {

    private val tags = arrayOf("uml", "salt", "math", "latex", "gantt")

    override fun compile(extension: String, source: String, data: Map<String, Any>?): String {
        var result = source
        try {
            try {
                // ensure string is wrapped in @startuml...@enduml
                result = result.trim { it <= ' ' }
                result = wrapDiagram(result)

                // compile string to SVG
                val os = ByteArrayOutputStream(1024)

                val fileFormat = FileFormatOption(FileFormat.SVG)
                fileFormat.hideMetadata()

                val reader = SourceStringReader(result)
                reader.outputImage(os, fileFormat)
                os.close()

                var s = String(os.toByteArray(), Charset.forName("UTF-8"))
                s = s.replace("<\\?(.*)\\?>".toRegex(), "") // remove XML declaration

                return s
            } catch (e: IOException) {
                e.printStackTrace()
                return ""
            }

        } catch (e: Throwable) {
            e.printStackTrace()
            return ""
        }
    }

    override fun getOutputExtension(): String {
        return "svg"
    }

    override fun getSourceExtensions(): Array<String> {
        return arrayOf("uml")
    }

    private fun wrapDiagram(input: String): String {
        var isWrapped = false

        var source = input

        for (tag in tags) {
            if (source.startsWith("@start" + tag)) {
                isWrapped = true
                break
            }
            if (source.endsWith("@end" + tag)) {
                isWrapped = true
                break
            }
        }

        if (!isWrapped) {
            source = "@startuml\n" + source
            source = source + "\n@enduml"
        }

        return source
    }
}
